File: pathsum2_backtrack.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    List<List<Integer>> result;
    int target;
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        
        result=new ArrayList<>();
        target=sum;
        
        List<Integer> path=new ArrayList<>();
        
        helper(root,0,path);
        
        return result;
    }
    
    private void helper(TreeNode root, int currSum, List<Integer> path)
    {
        if(root==null) return;
        
        /* Action */
        
        /* Check if its a leaf node */
        path.add(root.val);
        if(root.left==null && root.right==null)
        {
            if(currSum+root.val==target)
                result.add(new ArrayList<>(path));
        }
        
        /* Recurse */
        
        helper(root.left,currSum+root.val,path);
        helper(root.right,currSum+root.val,path);
        
        /* Backtrack */
            
        path.remove(path.size()-1);
    }
}

File: pathsum2_backtrack.py
'''
Backtrack Solution
Time Complexity : O(n)
Space Complexity: O(h)
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        out = []
        def sum1Backtrack(root, localSum,path):
            if(root==None):
                return
            localSum += root.val
            if(root.left==None and root.right==None):
                nonlocal out
                if(localSum == sum):
                    out.append(path + [root.val])
                return
            path.append(root.val)
            
            sum1Backtrack(root.left, localSum, path)
            sum1Backtrack(root.right, localSum, path)
            # BACKTRACKING
            path.pop()
        
        sum1Backtrack(root, 0, [])
        return out
